lib/errordedup/errordedup.go

```diff
@@ -0,0 +1,205 @@
+	timeWindow = time.Minute
+	// timeWindowCleanupInterval is the interval between cleanups of
+	// time windows that have already ended.
+	// Since the time window is set to 1 minute and the cleanup interval
+	// is set to 10 seconds, time windows can be in fact slightly larger
+	// than 1 minute (i.e., 1 minute and 10 seconds, in the worst case).
+	timeWindowCleanupInterval = 10 * time.Second
+)
+
+// Config contains the error deduplicator config.
+type Config struct {
+	// Entry is the logger entry.
+	Entry *log.Entry
+	// LogLevel is the log level at which errors will be logged.
+	// The default log level (0) is panic.
+	LogLevel log.Level
+	// DebugReport indicates whether a debug report should be generated
+	// for the errors reported.
+	DebugReport bool
+	// ErrorSubstrings contains a list of substrings belonging to the
+	// errors that should be deduplicated.
+	ErrorSubstrings []string
+	// ChannelSize is the size of the channel used to send messages
+	// to the error deduplicator.
+	ChannelSize int
+	// Clock is a clock to override in tests, set to real time clock
+	// by default.
+	Clock clockwork.Clock
+}
+
+// CheckAndSetDefaults verifies configuration and sets defaults
+func (c *Config) CheckAndSetDefaults() error {
+	if c.Entry == nil {
+		c.Entry = log.WithFields(log.Fields{
+			trace.Component: "errordedup",
+		})
+	}
+	if c.ErrorSubstrings == nil {
+		return trace.BadParameter("missing parameter ErrorMessages")
+	}
+	if c.ChannelSize < 0 {
+		return trace.BadParameter("ChannelSize must be at least 0")
+	}
+	if c.Clock == nil {
+		c.Clock = clockwork.NewRealClock()
+	}
+	return nil
+}
+
+// ErrorDeduplicator deduplicates errors over a certain time window.
+type ErrorDeduplicator struct {
+	Config
+	// errorCh is used to send errors to the deduplicator.
+	errorCh chan error
+	// errorMap contains deduplicated errors reported over
+	// certain time windows.
+	errorMap map[string]*errorWindowInfo
+}
+
+// errorWindowInfo contains the information necessary to deduplicate an error
+// reported within a certain time window.
+type errorWindowInfo struct {
+	// firstError is the first error reported within this window.
+	firstError string
+	// timeWindowStart is the beginning of the time window, i.e.,
+	// the time at which the error was first reported (for this window).
+	timeWindowStart time.Time
+	// occurrences are the occurrences of errors (that share the same
+	// error substring) within this window.
+	occurrences int
+}
+
+// New creates a new error deduplicator.
+func New(cfg Config) (*ErrorDeduplicator, error) {
+	if err := cfg.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &ErrorDeduplicator{
+		Config:   cfg,
+		errorCh:  make(chan error, cfg.ChannelSize),
+		errorMap: make(map[string]*errorWindowInfo, len(cfg.ErrorSubstrings)),
+	}, nil
+}
+
+// Send sends an error to the error deduplicator.
+func (e *ErrorDeduplicator) Send(err error) {
```

Comment:
I think the design here can be simpler: we don't need channels, goroutines, an implicit log level or even to hold a logger.

To begin, consider a signature like this:

```go
func (l *LogLimiter) Log(level log.Level, entry *log.Entry)
```

The `LogLimiter` keeps a timestamped, sorted listed of entries. When `Log` is called it trims entries that are too old, looks for dupes, then logs/records the new entry as applicable. There's no need for concurrency and the algorithm is quiescent, unless called.

It seems the only data we might need are the strings to match, but I'm not sure that belongs here either - I think the caller should choose whether to use the limiter, and the limiter keeps every string it sees for the time window. This is likely what we want, anyway.

Reply:
Thanks for the review @codingllama!

> I think the design here can be simpler: we don't need channels, goroutines, an implicit log level or even to hold a logger.

I believe that we need concurrency because `Send`/`Log` can be called from multiple goroutines (`Mux.detectAndForward` in this PR), and we want to deduplicate/rate limit over all of them, not per goroutine.

Regarding the implicit log level, we could change this method to be e.g.:

```suggestion
func (e *ErrorDeduplicator) Log(level log.Level, err error) {
```
---
> The LogLimiter keeps a timestamped, sorted listed of entries.

Won't this be less performant than an hash map? At least for inserts (`deduplicate`s in this PR) which should be must more frequent than `cleanup`s.

---
> There's no need for concurrency and the algorithm is quiescent, unless called.

If the algorithm is quiescent, then I'm not sure what we're supposed to do when the 3rd error below is logged:
- `"A error 1"` at `10:00:00`: the error is logged
- `"A error 2"` at `10:00:01`: the error is not logged
- `"A error 3"` at `10:30:00`: the error is logged and we say that 1 or 2 similar logs happened in the past 30m?

Reply:
> I believe that we need concurrency because Send/Log can be called from multiple goroutines

Reply:
> I believe that we need concurrency because Send/Log can be called from multiple goroutines (...)

A mutex can handle that.

> Won't this be less performant than an hash map

Adding an cleaning up are optimal in a scheme like this, traversal is essentially the same. (Unless I'm missing something?)

> "A error 3" at 10:30:00: the error is logged and we say that 1 or 2 similar logs happened in the past 30m?

How big is the time window? If it's 1m, then `10:00:00` and `10:00:01` are removed once the `10:30:00` call happens, so it gets logged. We might hold on to memory for a long time if there are no errors, but if the lists aren't big then it won't matter much (otherwise we can remedy just that).

Reply:
> I think the design here can be simpler: we don't need channels, goroutines, an implicit log level or even to hold a logger.

Removed the implicit log level and logger in 300852b06bbf68c87996bb3c794f1c5653e00ca4. The API is now:

```go
func (l *LogLimiter) Log(entry *log.Entry, level log.Level, args ...any) {
```

Reply:
Looks better. I still think there's a version of this that doesn't need channels or goroutines, but let's move forward.